website/api/controllers/issue-est-certificate.js

```diff
@@ -0,0 +1,136 @@
+    const INTROSPECT_ENDPOINT = sails.config.custom.certIssueIdpIntrospectEndpoint;
+    if (!INTROSPECT_ENDPOINT) {
+      throw new Error('sails.config.custom.certIssueIdpIntrospectEndpoint is required');
+    }
+    const IDP_CLIENT_ID = sails.config.custom.certIssueIdpClientId;
+    if (!IDP_CLIENT_ID) {
+      throw new Error('sails.config.custom.certIssueIdpClientId is required');
+    }
+    const EST_ENDPOINT = sails.config.custom.certIssueEstEndpoint;
+    if (!EST_ENDPOINT) {
+      throw new Error('sails.config.custom.certIssueEstEndpoint is required');
+    }
+    const EST_CLIENT_ID = sails.config.custom.certIssueEstClientId;
+    if (!EST_CLIENT_ID) {
+      throw new Error('sails.config.custom.certIssueEstClientId is required');
+    }
+    const EST_CLIENT_KEY = sails.config.custom.certIssueEstClientKey;
+    if (!EST_CLIENT_KEY) {
+      throw new Error('sails.config.custom.certIssueEstClientKey is required');
+    }
+
+    // Ask the IdP to introspect the auth token (ensure it's valid and extract the values).
+    const introspectResponse = await sails.helpers.http.sendHttpRequest.with({
+      url: INTROSPECT_ENDPOINT,
+      method: 'POST',
+      enctype: 'application/x-www-form-urlencoded',
+      body: {
+        'client_id': IDP_CLIENT_ID,
+        'token': authToken,
+      },
+    });
+
+    if (!introspectResponse.body) {
+      throw 'invalidToken';
+    }
+
+    if (!introspectResponse.body.active) {
+      throw 'invalidToken';
+    }
+
+    const introspectUsername = introspectResponse.body.username;
+
+    // Extract the email and username from the CSR. Ensure they match.
+    let jsrsasign = require('jsrsasign');
+    const csrUtil = jsrsasign.asn1.csr.CSRUtil;
+    const csrObj = csrUtil.getParam(csrData);
+    let csrEmail = '';
+    let csrUsername = '';
+    for (const extension of csrObj.extreq) {
+      if (extension.extname === 'subjectAltName') {
+        for (const extentry of extension.array) {
+          if ('rfc822' in extentry) {
+            csrEmail = extentry.rfc822;
+          }
+          if ('other' in extentry) {
+            if ('oid' in extentry.other && extentry.other.oid === '1.3.6.1.4.1.311.20.2.3') {
+              csrUsername = extentry.other.value.utf8str.str;
+            }
+          }
+        }
+      }
+    }
+    if (csrEmail === '') {
+      throw 'badRequest';
+    }
+    if (!csrEmail.startsWith(csrUsername)) {
+      throw 'badRequest';
+    }
+
+    // Ensure username from IdP auth matches username in CSR. If they don't match, perhaps the user
+    // is trying to get a certificate with another user's name?
+    if (csrEmail !== introspectUsername) {
+      throw 'invalidToken';
+    }
+
+    // Ask the PKI provider for a certificate
+    const estResponse = await sails.helpers.http.sendHttpRequest.with({
+      url: EST_ENDPOINT,
+      method: 'POST',
+      body: csrData.replace(/(-----(BEGIN|END) CERTIFICATE REQUEST-----|\n)/g, ''),
+      enctype: 'text/plain',
```

Comment:
@eashaw unfortunately this doesn't seem to work even though `text/plain` is a valid `enctype`.

Are there docs for this stuff? AFAICT it seems to use [this library](https://www.npmjs.com/package/machinepack-http) and looking in the code it seems like there may be no way to just send a string as a request body?  https://github.com/sailshq/machinepack-http/blob/master/lib/send-http-request.js#L68-L74

Does a typical sails user just know that this library is used under the hood and then read the code to understand what the options are?

Reply:
The API we are working with here requires that the csr data is the only thing in the request body and that the `Content-Type` header is `application/pkcs10`.

Reply:
I could also just revert the change and continue to use axios which was working as expected. What's your preference?

Reply:
@zwass Sending a `text/plain` or `application/pkcs10` encoded request is not possible with the built-in sendHttpRequest helper. We may need to revert back to using axios or (preferably) use the request package that the sails helper uses under the hood for this request. 
Depending on how time-sensitive this PR is, we could also add support for those enctypes to that helper and publish a new version of `machinepack-http`.

> Does a typical sails user just know that this library is used under the hood and then read the code to understand what the options are?

I'm not sure! To see all available options, I usually look at the source code, or run the website in an interactive mode by starting the server with `sails console` and run the getDef() method on helpers to see all options in the terminal. (e.g., running `sails.helpers.http.sendHttpRequest.getDef()`)

Reply:
@eashaw it was actually an easy change to use the underlying request library: https://github.com/fleetdm/fleet/pull/23637/commits/cd807ef7cdff764e3d27c6c54fd584572531c1eb